### PR TITLE
Added the DaccredBase contract.

### LINKED ISSUE #12: DaccredBase.sol
This would be the core implementation of a daccred. 

`DaccredBase.sol`  is an ERC-721 token that adopts some key structures from `ERC721A` 
Once we have this, we will then need another Factory that can create child `Core` contracts. will document the specs for that in another issue. 

However the scope of this issue is an NFT contract that supposed the full specification of the ERC-721 but is gas optimized and cost very little to mint like the implementation https://github.com/chiru-labs/ERC721A

### PR DESCRIPTION
**Pull request by 0xfps**
# 6
DaccredBase.all

Fixes #12 

## Added the DaccredBase contract.

Added the DaccredBase contract, required according to issue #12 .

Environment

| Environment                                       | Version           |
| ------------------------------------------------- | ----------------- |
| Compiler version                                  |  0.8.7   |
| Target EVM version (as per compiler settings):    |  default  |
| Framework/IDE (e.g. Truffle or Remix)             |  Remix    |
| Operating system (macOS / Windows)                |  Windows   |

### Description

The complete implementation is the DaccredBase contract was added, this will be deployed by the Factory when written.

### CODE REVIEW

**Comment by koolamusic** [comment]
this is gonna take a while

**Comment by 0xfps** [comment]
Ok.

File: src/contracts/DaccredBase.sol [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+    *                   msg.sender.
+    */
+    constructor(
+        string memory _name, 
+        string memory _symbol, 
+        address _owner
+    )
+    {
+        /// @dev Assign name.
+        tokenName = _name;
```
Comment (koolamusic) [inline src/contracts/DaccredBase.sol]: this is a problem.
Variable names should be exact with the ERC specification. 

Ideally 
symbol
name
burnCounter
indext
tokenIndex
currentIndex


these are common variable names used across the 721 spec

File: src/contracts/DaccredBase.sol [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+    string private tokenName;
+    /// @dev Token symbol.
+    string private tokenSymbol;
+    /// @dev Owner of the tokens.
+    address private owner;
+    /// @dev Stores the number of tokens minted.
+    /// @dev Starts from 0 and is incremented per mint based on quantity minted.
+    uint256 private tokensMinted;
+    /// @dev Stores the number of tokens burned.
+    uint256 private tokensBurned;
```
Comment (koolamusic) [inline src/contracts/DaccredBase.sol]: review these names

**Review by koolamusic — CHANGES_REQUESTED** [review]
left a few comments, however in order to be explicit. 

Let's copy this contract https://github.com/chiru-labs/ERC721A/blob/main/contracts/ERC721A.sol

about 95% of the definitions of the ERC721A should be the ERC721 Extension.
Base contract should inherit it with any other contracts we use to add auth or utility to `DaccredBase`

**Comment by 0xfps** [comment]
> left a few comments, however in order to be explicit.
> 
> Let's copy this contract https://github.com/chiru-labs/ERC721A/blob/main/contracts/ERC721A.sol
> 
> about 95% of the definitions of the ERC721A should be the ERC721 Extension. Base contract should inherit it with any other contracts we use to add auth or utility to `DaccredBase`

I have effected the changes 😊.

**Comment by koolamusic** [comment]
LGTM

**Pull request merged**